pyro/ops/integrator.py

```diff
@@ -44,3 +34,38 @@ def _grad(potential_fn, z):
     grads = grad(potential_fn(z), z_nodes)
     grads = [v.data for v in grads]
     return dict(zip(z_keys, grads))
+
+
+def velocity_verlet(z, r, potential_fn, step_size, num_steps, transforms=None):
+    """
+    Second order symplectic integrator that uses the velocity verlet algorithm.
+
+    :param dict z: dictionary of sample site names and their current values
+        (type ``torch.autograd.Variable``).
+    :param dict r: dictionary of sample site names and corresponding momenta
+        (type ``torch.autograd.Variable``).
+    :param callable potential_fn: function that returns potential energy given z
+        for each sample site. The negative gradient of the function with respect
+        to ``z`` determines the rate of change of the corresponding sites'
+        momenta ``r``.
+    :param float step_size: step size for each time step iteration.
+    :param int num_steps: number of discrete time steps over which to integrate.
+    :param dict transforms: Optional dictionary that specifies a transform
+        for a sample site with constrained support to unconstrained space. The
+        transform should be invertible, and implement `log_abs_det_jacobian`.
+    :return tuple (z_next, r_next): final position and momenta, having same types as (z, r).
+    """
+    if transforms is None:
+        return unconstrained_velocity_verlet(z, r, potential_fn, step_size=step_size, num_steps=num_steps)
+
+    def unconstrained_potential_fn(u):
+        z = {name: transforms.get(name, identity_transform).inv(u_name) for name, u_name in u.items()}
```

Comment:
Do you think it's worth eliminating the dependency on `identity_transform` by expanding our logic a bit here and below?
```py
z = u.copy()
for name, transform in transforms.items():
    z[name] = transform(z[name])
```

Reply:
Unless we have some perf considerations (both seem like they should be about the same), I like the pattern of using the default identity transform above. We should change `u_name` to `unconstrained_value` or just `value` to make it clear that it is a tensor object.